VDiff v2 Blog Post (#1219)
vitessio/website: content/en/blog/2022-11-14-vdiff-v2.md

---
author: 'Matt Lord'
date: 2022-11-14
slug: '2022-11-14-vdiff-v2'
tags: [Vitess','MySQL','sharding','replication']
title: 'Introducing Vitess Diff V2'

> description: "Vitess's Powerful Data Diff Tool Just Got Even Better"

Reviewer: I had some meta questions about the blog post. Should we be focusing more on VDiff itself in the blog post? Currently we are focussing on why V2 is better. 

While that is interesting to existing users who have faced issues with v1, IMHO the blog post will have more appeal if it we go forward logically from the `Why a Diff Tool` section to explaining the challenges of doing a diff on a VReplication workflow: both due to large data sizes and the fact that we are doing it on a running workflow and an online cluster. And then elucidating the VDiff algorithm.

Also, prima facie, the impression I get is that VDiff  is a generic data diff tool: we should probably make it clearer it only works for existing VReplication workflows and not to compare any two tables ...

I will let @deepthi also weigh in. If we want to stay with the current format I will review further.

Author: We had one of 4 sections (and it was the smallest one) on why VDiff2 was an improvement. 🙂 

But I've incorporated your feedback -- insomuch as it made sense to me when considering the full post -- here: https://deploy-preview-1219--vitess.netlify.app/blog/2022-11-14-vdiff-v2/#vdiff

Reviewer 2: I see @rohit-nayak-ps said
> Also, prima facie, the impression I get is that VDiff is a generic data diff tool: we should probably make it clearer it only works for existing VReplication workflows and not to compare any two tables ...

which was also my first comment on the title. So yes, let's make it super clear it's only applicable for VRepl workflows.

---
author: 'Matt Lord'
date: 2022-11-14
slug: '2022-11-14-vdiff-v2'
tags: [Vitess','MySQL','sharding','replication']

> title: 'Introducing Vitess Diff V2'

Reviewer 2: Technically it's VReplication Diff, not Vitess Diff.
You can't diff arbitrary keyspaces against each other can you? You can only diff the source and target of a VReplication workflow.
I would just say `Introducing VDiff V2`.

Author: vdiff, vschema, vindex, vgtid etc AFAIUI is vitess &lt;blank&gt;. But I do agree that I think to be more precise it would be VReplication VDiff: https://github.com/vitessio/vitess/pull/5367 

I'll use that when longer form may be nicer and agree with you on VDiff for the title as VReplication [V|Vitess ]Diff is a bit much... :)

description: "Vitess's Powerful Data Diff Tool Just Got Even Better"
---

Vitess is a solution that allows you to infinitely scale MySQL while providing clients and apps with a single logical

> mysqld view of the fleet of MySQL instances comprising any number of [`Keyspaces`](https://vitess.io/docs/concepts/keyspace/)

Reviewer 2: ```suggestion 
view of the fleet of MySQL instances comprising any number of [`Keyspaces`](https://vitess.io/docs/concepts/keyspace/)
```

  * [lock the table](https://dev.mysql.com/doc/refman/en/lock-tables.html) to get the current
    [`GTID_EXECUTED`](https://dev.mysql.com/doc/refman/en/replication-gtids-concepts.html) which gives us a logical
    point in time that will correspond to the the read view in our upcoming transaction
  * issue [`START TRANSACTION WITH CONSISTENT SNAPSHOT`](https://dev.mysql.com/doc/refman/en/commit.html)
  * [unlock the table](https://dev.mysql.com/doc/refman/en/lock-tables.html) as we now have a consistent snapshot of
    the table data and the GTID metadata that are both at the same logical point in time with regards to the table
    we're diffing
* On the target [tablets](https://vitess.io/docs/concepts/tablet/):
  * start VReplication UNTIL we have reached that `GTID_EXECUTED` position in the [VStream](https://vitess.io/docs/concepts/vstream/)
    which matches the one we saved when setting up the read view on the source
  * issue [`START TRANSACTION WITH CONSISTENT SNAPSHOT`](https://dev.mysql.com/doc/refman/en/commit.html) (remember
    that the state is "frozen" on the target tablet) — now the target context is at the same logical point in
    time as the source for this table

> * On the source and target tablets: issue `SELECT <cols>, FROM <table> ORDER BY <pkcols>` (for deterministic ordering and to avoid a filesort)

Reviewer: Should we remove the comma after `<cols>`?

Author: Good catch! Removed.

---
author: 'Matt Lord'
date: 2022-11-14
slug: '2022-11-14-vdiff-v2'
tags: [Vitess','MySQL','sharding']
title: 'Vitess Diff V2'
description: "Vitess's Powerful Data Diff Tool Just Got Better"
---

> Vitess is a solution that allows you to infinitely scale MySQL while providing clients and apps with a single mysqld

Reviewer 3: Personal wording preference
```suggestion
Vitess is a solution that allows you to infinitely scale MySQL while providing applications with a single logical
```

> view of the fleet of MySQL instances that are managing N [Keyspaces](https://vitess.io/docs/concepts/keyspace/) and

Reviewer 3: ```suggestion
view of the fleet of MySQL instances comprising one or more [Keyspaces](https://vitess.io/docs/concepts/keyspace/) and
```

> Y [Shards](https://vitess.io/docs/concepts/shard/).

Reviewer 3: ```suggestion
one or more [Shards](https://vitess.io/docs/concepts/shard/).
```

<img src="/img/VitessQueryExample.png" alt="Query Example" width="275" align="right"/>

Vitess also provides the cluster and data management tools that make it possible to manage a massive cluster and
perform complex workflows for [moving tables](https://vitess.io/docs/reference/vreplication/movetables/),

> [resharding](https://vitess.io/docs/reference/vreplication/reshard/) to adjust to changes in data size and load,

Reviewer 3: This sentence is pretty long and unwieldy. Might help to break up into multiple sentences, or maybe turn everything after `perform complex workflows for` into bullet points.

[materialized views and rollups](https://vitess.io/docs/reference/vreplication/materialize/) for data analytics
and data locality, performing [online schema changes](https://vitess.io/docs/user-guides/schema-changes/managed-online-schema-changes/),
and more using [VReplication](https://vitess.io/docs/16.0/reference/vreplication/vreplication/).

## Why VDiff?

Data is typically one of the most critical assets within an organization. As such, an operator needs to be able to

> verify the correctness of this data, in particular as the data is moved around or otherwise transformed. Even for

Reviewer 3: This word feels more natural to me here, but maybe just a personal preference.

```suggestion
verify the correctness of this data, especially as the data is moved around or otherwise transformed. Even for
```

> the relatively simple and reliable case of replicating data from one MySQL instance to another there's been a

Reviewer 3: You have two sentence in a row starting with `even for`. I'd try a different setup:

> For example, operators want  a way to verify data consistency after replicating data from one MySQL instance to another. However, even for a single table...

strong desire for some kind of easy-to-use verification tool. But even for a single table in this simplest of case — 
performing a safe, reliable, light-weight, and performant diff between two MySQL instances is a suprisingly
difficult problem. Due to the challenges involved, there have been few attempted general solutions with the most
notable being:
  * Percona's [pt-table-checksum](https://docs.percona.com/percona-toolkit/pt-table-checksum.html)
  * MySQL's `mysqldiff` tool that was part of the now EOL'd [MySQL Utilities](https://downloads.mysql.com/docs/mysql-utilities-1.6-en.pdf)

> With Vitess, _the need for a data diff tool is even more pronounced_ because you'll be migrating data from your legacy systems into Vitess, migrating data across different sharding schemes, and performing a host of various data transformations/materializations.  This is further complicated by the fact that these migarations may be done across MySQL versions, data centers, and with differing schemas on the source and target as your data evolves over time and is modified to e.g. support new sharding or data distribution schemes. So the need for a tool that can reliably perform a logical diff between the source and target of these data migrations and workflows — which also likely have different data sharding/partitioning schemes in place — all in a timely manner and without impacting production traffic regardless of data set size and other complicating factors becomes critical for peace of mind as you move your most valuable assets around.

Reviewer 3: Nit: with the previous parts of the blog post there's a shorter line limit than with this line. Realize that will be rendered a away by Markdown, but making a note of it anyway.

This paragraph is a bit unwieldy. Tried to simplify a tiny bit.

```suggestion
With Vitess, _the need for a data diff tool is even more pronounced_ because you'll be migrating data from your legacy systems into Vitess, migrating data across keyspaces, and performing a variety of other workflows. This is further complicated by the fact that these workflows may be done across MySQL versions, data centers, with differing schemas between the source and target, and over long timer periods in which your data evolves. So, it is critical to have a tool that can reliably perform a logical diff between the source and target of these workflows, in a timely manner, and without impacting production traffic.
```
